File: wiki_java/src/main/java/com/github/klee0kai/wiki/wrapping/CustomWrapper.java
package com.github.klee0kai.wiki.wrapping;

import com.github.klee0kai.stone.annotations.wrappers.WrappersCreator;
import com.github.klee0kai.stone.wrappers.creators.ProviderWrapper;

import javax.inject.Provider;

@WrappersCreator(wrappers = {CustomLazy.class})
public class CustomWrapper implements ProviderWrapper {

    @Override
    public <Wr, T> Wr wrap(Class<Wr> wrapperCl, Provider<T> originalProvider) {
        if (wrapperCl.equals(CustomLazy.class)) {
            return (Wr) new CustomLazy<T>(originalProvider::get);
        }
        return null;
    }

}

File: wiki_kotlin/src/main/kotlin/com/github/klee0kai/wiki/wrapping/CustomWrapper.kt
package com.github.klee0kai.wiki.wrapping

import com.github.klee0kai.stone.annotations.wrappers.WrappersCreator
import com.github.klee0kai.stone.wrappers.creators.ProviderWrapper
import javax.inject.Provider

@WrappersCreator(wrappers = [CustomLazy::class])
class CustomWrapper : ProviderWrapper {

    override fun <Wr, T> wrap(wrapperCl: Class<Wr>, originalProvider: Provider<T>): Wr? {
        if (wrapperCl == CustomLazy::class.java) {
            return CustomLazy { originalProvider.get() } as Wr
        }
        return null
    }

}
